addons/delivery_provider/models/delivery_carrier.py

```diff
@@ -0,0 +1,140 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import logging
+from openerp import models, fields, api
+from openerp.exceptions import UserError
+
+_logger = logging.getLogger(__name__)
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = 'delivery.carrier'
+    ''' A Shipping Provider
+
+    In order to add your own external provider, follow these steps:
+
+    1. Create your model MyProvider that _inherit 'delivery.carrier'
+    2. Extend the selection of the field "delivery_type" with a pair
+       ('<my_provider>', 'My Provider')
+    3. Add your methods:
+       <my_provider>_get_shipping_price_from_so
+       <my_provider>_send_shipping
+       <my_provider>_open_tracking_page
+       <my_provider>_cancel_shipment
+       (they are documented hereunder)
+    '''
+
+    ####################################
+    # Internals for shipping providers #
+    ####################################
+
+    delivery_type = fields.Selection([('default', 'Default Delivery')], string="Delivery Type", default='default', required='True')
+
+    price = fields.Float(compute='get_price')
+    available = fields.Boolean(compute='get_price')
+
+    @api.one
+    def get_price(self):
+        SaleOrder = self.env['sale.order']
+
+        self.available = False
+        self.price = False
+
+        order_id = self.env.context.get('order_id')
+        if order_id:
+            if 'default' not in self.delivery_type:
+                try:
+                    order = SaleOrder.browse(order_id)
+                    self.price = self.get_shipping_price_from_so(order)[0]
+                    self.available = True
+                except UserError as e:
+                        # no suitable delivery method found, probably configuration error
+                        _logger.info("Carrier %s: %s, not found", self.name, e.name)
+                        self.price = 0.0
+            else:
+                res = super(DeliveryCarrier, self).get_price('price', [])
+                self.available = res[self.id]['available']
+                self.price = res[self.id]['price']
+
+    #########################################
+    # Hooks to implement external providers #
+    #########################################
+
+    # TODO definir et gerer les exceptions qui peuvent etre lancees par ces providers
+
+    def get_shipping_price_from_so(self, orders):
```

Comment:
@KangOl But doing this will encapsulate every result in a list containing a list of results... I'm not sure, but in fact, these methods are more multi (and by the way they are already called on singletons)... I'd suggest a self.ensure_one to avoid ugly unpackings